### ISSUE TITLE
Variable type is narrowed incorrectly when it is assigned later within a loop

### ISSUE THREAD

**Issue by rpmrmartin** [microsoft/pyright#6075]
**Describe the bug**
1. Declare two `Optional` variables and initialize them to `None`
2. Inside a loop, compare each variable to a non-optional object attribute of the same type. Here, Pyright will report an unnecessary comparison on the second comparison, but not the first.
3. Later in the loop, conditionally assign each variable a non-optional value

**Code or Screenshots**
```python
# pyright: strict

from dataclasses import dataclass

@dataclass
class Clazz:
	Prop1: str
	Prop2: str

iterator = iter([Clazz('a', 'b')])
strOrNone: str | None = None
strOrNone2: str | None = None
while True:
	try:
		obj = next(iterator)
	except StopIteration:
		obj = None

	condition = obj is not None and (obj.Prop1 != strOrNone or obj.Prop2 != strOrNone2) # reportUnnecessaryComparison on this line

	if obj is not None:
		strOrNone = obj.Prop1
		strOrNone2 = obj.Prop2
	else:
		break
```

**VS Code extension or command-line**
VS Code Python extension v2023.16.0
CLI v1.1.329

**Comment by erictraut** [comment]
Thanks for the bug report. I'm able to repro. 

I was able to pare it down to a simpler example:

```python
# pyright: strict
def func1() -> str | None:
    ...

s1: str | None = None
s2 = None
while True:
    obj = func1()

    x = s2
    condition = obj and obj != s1

    s1 = obj
    s2 = obj
```

**Comment by erictraut** [comment]
Thanks for the bug fix. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.331, which I just published. It will also be included in a future release of pylance.
